### ISSUE TITLE
[PEP] Rename and document `pyomo.contrib`

### ISSUE THREAD

**Issue by mrmundt** [Pyomo/pyomo#2460]
## Summary

There have long been discussions surrounding the name of `pyomo.contrib` - primarily influenced by the desire to make a more appropriately and clearly designated distinction between the main development team's core, supported features and those which are actively under development/"use at your own risk."

### Rationale

The original intention behind `pyomo.contrib` was to be a location in which to place third-party contributions. While this is still true, the waters have been somewhat muddied regarding support level of these items. We'd like to rename the directory in order to bring clarity to the true intended purpose of this directory - that is, a place where active development and "use at your own risk" features are located.

### Proposed Solutions

1. `pyomo.contrib` -> `pyomo.sandbox`
2. `pyomo.contrib` -> `pyomo.devel`

In either case, we want to add clarifying documentation in the new directory as to the purpose which will include: contribution guide, requirements on contributions (e.g., the fact that they are less strict than re-architecting of existing core functionality), etc. This documentation will require some iteration and will need review by the PMC.

**Comment by bernalde** [comment]
I totally agree with this! My vote would be for pyomo.devel (sounds more serious than sandbox and implicitly encodes the active development part of the description).

**Comment by emma58** [comment]
I also agree with this. I prefer sandbox because I'd rather emphasize that it's experimental and that there is little initial commitment to what we put in there--we might tear down the sand castle and start over. But devel would be fine too.

**Comment by mrmundt** [comment]
The newest idea here:

- `dev`: a place where things are actively under development / no promise of stability
   - MindtPy, PyROS, Coramin, LaTeX printer
- `apps`: a place to store functionality that is not a core offering but has some level of promised stability
   -  PyNumero, FME, TRF
- `supported`: the main offering; the place to which apps get promoted; everything that is "top-level" and is actively supported
   - FBBT (**probably should go into core**, is currently in `contrib`)

In this way, it's a clear progression of stability from `devel` -> `apps` -> `supported`. This way we can more adequately organize into an appropriate structure without a focus on "where did this come from" and a stronger promise of stability at different levels.

UPDATE (Aug 12, 2025): `dev` -> `devel`; `apps` -> `addons`

**Comment by Robbybp** [comment]
In this framework, is there a difference between extensions in `supported` and "top-level" extensions such as DAE, GDP, and Network? Or are you proposing to combine these somehow?

**Comment by jsiirola** [comment]
@Robbybp: The proposal is (I think) *not* for a `supported` directory, but rather that it is a "category" that covers everything not in `dev` or `apps`.

**Comment by mrmundt** [comment]
A question has come up about dependency management - if something is moved to `supported`, what does that mean about the dependencies? Historically, there are things in `core` that use optional dependencies, but those deps are optional. We should probably follow that standard unless there is a strong reason otherwise. (We like our single "required" dep)

**Comment by mrmundt** [comment]
There is another question here about minimum expectations for each category. (I have already started working on this by looking at the existing contrib packages.)

**Comment by mrmundt** [comment]
Another wrench: if a package wants to do a rewrite, what happens? Does the rewrite become a parallel package that lives in `devel` and then eventually replaces the version in `addons`? (If it's a huge rewrite, yes.) It can be in-place if it is a less disruptive change.
